File: test_app/integration_tests/android_mocks/src/main/java/com/voxeet/asserts/NotificationServiceAsserts.java
package com.voxeet.asserts;

import android.util.Log;

import androidx.annotation.NonNull;

import com.voxeet.VoxeetSDK;
import com.voxeet.sdk.json.ParticipantInvited;
import com.voxeet.sdk.models.Conference;
import com.voxeet.sdk.push.center.subscription.register.BaseSubscription;
import com.voxeet.sdk.services.ConferenceService;
import com.voxeet.sdk.services.NotificationService;

import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import io.dolby.asserts.AssertUtils;
import io.dolby.asserts.MethodDelegate;

public class NotificationServiceAsserts implements MethodDelegate {
    @Override
    public void onAction(String methodName, Map<String, Object> args, MethodDelegate.Result result) {
        try {
            switch (methodName) {
                case "assertInviteArgs":
                    assertInviteArgs(args);
                    break;
                case "assertDeclineArgs":
                    assertDeclineArgs(args);
                    break;
                case "assertSubscribeArgs":
                    assertSubscribeArgs(args);
                    break;
                case "assertUnsubscribeArgs":
                    assertUnsubscribeArgs(args);
                    break;
                default:
                    result.error(new NoSuchMethodError());
                    return;
            }
        } catch (MethodDelegate.AssertionFailed exception) {
            //result.failed(exception);
        } catch (Exception ex) {
            result.error( ex);
        }
        result.success();
    }

    @Override
    public String getName() {
        return "IntegrationTesting.NotificationServiceAsserts";
    }

    private void assertInviteArgs(Map<String, Object> args) throws AssertionFailed, KeyNotFoundException{
        Object mockHasRun = VoxeetSDK.notification().inviteHasRun;
        if (!args.containsKey("hasRun")) {
            throw new KeyNotFoundException("Key: hasRun not found");
        } else {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "hasRun is incorrect");
        }
        NotificationService.InviteArgs mockArgs= VoxeetSDK.notification().getInviteArgs();
        if (!args.containsKey("conference")) {
            throw new KeyNotFoundException("Key: conference not found");
        } else {
            ConferenceServiceAsserts.assertConference((Map<String, Object>) args.get("conference"), mockArgs.getConference());
        }

        ParticipantInvited mockParticipantInvited = mockArgs.getParticipantsInvited().get(0);
        if (!args.containsKey("participantInvited")) {
            throw new KeyNotFoundException("Key: participantInvited not found");
        } else {
            Map<String, Object> participantInvited = (Map<String, Object>) args.get("participantInvited");
            if (!participantInvited.containsKey("name")) {
                throw new KeyNotFoundException("Key: name not found");
            } else {
                AssertUtils.compareWithExpectedValue(mockParticipantInvited.getParticipant().getName(), participantInvited.get("name"), "name is incorrect");
            }
            if (!participantInvited.containsKey("avatarUrl")) {
                throw new KeyNotFoundException("Key: avatarUrl not found");
            } else {
                AssertUtils.compareWithExpectedValue(mockParticipantInvited.getParticipant().getAvatarUrl(), participantInvited.get("avatarUrl"), "avatarUrl is incorrect");
            }
            if (!participantInvited.containsKey("externalId")) {
                throw new KeyNotFoundException("Key: externalId not found");
            } else {
                AssertUtils.compareWithExpectedValue(mockParticipantInvited.getParticipant().getExternalId(), participantInvited.get("externalId"), "externalId is incorrect");
            }
        }
    }

    private void assertDeclineArgs(Map<String, Object> args) throws AssertionFailed, KeyNotFoundException{
        Object mockHasRun = VoxeetSDK.notification().declineHasRun;
        if (!args.containsKey("hasRun")) {
            throw new KeyNotFoundException("Key: hasRun not found");
        } else {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "hasRun is incorrect");
        }

        Conference mockArgs = VoxeetSDK.notification().declineArgs;
        if (!args.containsKey("conference")) {
            throw new KeyNotFoundException("Key: conference not found");
        } else {
            ConferenceServiceAsserts.assertConference((Map<String, Object>) args.get("conference"), mockArgs);
        }
    }

    private void assertSubscribeArgs(Map<String, Object> args) throws AssertionFailed, KeyNotFoundException{
        boolean mockHasRun = VoxeetSDK.notification().subscribeHasRun;
        if (!args.containsKey("hasRun")) {
            throw new KeyNotFoundException("Key: hasRun not found");
        } else {
            AssertUtils.compareWithExpectedValue(mockHasRun, (boolean) args.get("hasRun"), "hasRun is incorrect");
        }

        List<BaseSubscription> mockArgs = VoxeetSDK.notification().subscribeArgs;
        String conferenceAlias = VoxeetSDK.conference().getConference().getAlias();
        if (!args.containsKey("subscription")) {
            throw new KeyNotFoundException("Key: subscriptions not found");
        } else {
            Map<String, Object> subscription = (Map<String, Object>) args.get("subscription");
            if (!subscription.containsKey("type")) {
                throw new KeyNotFoundException("Key: type not found");
            } else {
                AssertUtils.compareWithExpectedValue(mockArgs.get(0).type, subscription.get("type") , "type is incorrect");
            }
            if (!subscription.containsKey("conferenceAlias")) {
                throw new KeyNotFoundException("Key: conferenceAlias not found");
            } else {
                AssertUtils.compareWithExpectedValue(conferenceAlias, subscription.get("conferenceAlias") , "conferenceAlias is incorrect");
            }
        }
    }

    private void assertUnsubscribeArgs(Map<String, Object> args) throws AssertionFailed, KeyNotFoundException{
        Object mockHasRun = VoxeetSDK.notification().unsubscribeHasRun;
        if (!args.containsKey("hasRun")) {
            throw new KeyNotFoundException("Key: hasRun not found");
        } else {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "hasRun is incorrect");
        }

        List<BaseSubscription> mockArgs = VoxeetSDK.notification().unsubscribeArgs;
        String conferenceAlias = VoxeetSDK.conference().getConference().getAlias();
        if (!args.containsKey("subscription")) {
            throw new KeyNotFoundException("Key: subscriptions not found");
        } else {
            Map<String, Object> subscription = (Map<String, Object>) args.get("subscription");
            if (!subscription.containsKey("type")) {
                throw new KeyNotFoundException("Key: type not found");
            } else {
                AssertUtils.compareWithExpectedValue(mockArgs.get(0).type, subscription.get("type") , "type is incorrect");
            }
            if (!subscription.containsKey("conferenceAlias")) {
                throw new KeyNotFoundException("Key: conferenceAlias not found");
            } else {
                AssertUtils.compareWithExpectedValue(conferenceAlias, subscription.get("conferenceAlias") , "conferenceAlias is incorrect");
            }
        }
    }
}

File: test_app/integration_tests/ios_mocks/VoxeetSDKMock/Classes/NotificationServiceAsserts.swift
import Foundation

public class NotificationServiceAsserts {

    private static var instance: NotificationServiceAsserts?
    public static func create() {
        instance = .init()
    }

    var methodChannel: AssertsMethodChannel?

    public init() {
        assertsMethodChannelFactory.createAssertMethodChannel(
            forSdkAsserts: self,
            channelName: "IntegrationTesting.NotificationServiceAsserts"
        )
    }

    private func assertSubscribeArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.notification.declineHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }

        let mockArgs = VoxeetSDK.shared.notification.declineArgs
        try ifKeyExists(arg: args, key: "conference") { args in
            try ConferenceServiceAssertUtils.assertConference(
                args: args, mockConference: mockArgs
            )
        }
    }

    private func assertUnsubscribeArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.notification.declineHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }

        let mockArgs = VoxeetSDK.shared.notification.declineArgs
        try ifKeyExists(arg: args, key: "conference") { args in
            try ConferenceServiceAssertUtils.assertConference(
                args: args, mockConference: mockArgs
            )
        }
    }

    private func assertInviteArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.notification.inviteHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }

        let mockArgs = VoxeetSDK.shared.notification.inviteArgs
        try ifKeyExists(arg: args, key: "conference") { conference in
            try ConferenceServiceAssertUtils.assertConference(
                args: conference, mockConference: mockArgs?.conference
            )
        }

        let mockParticipantInvited = mockArgs?.participantsInvited.first
        try ifKeyExists(arg: args, key: "participantInvited") { (participantInvited: [String:Any]) in
            try ifKeyExists(arg: participantInvited, key: "name") { name in
                try nativeAssertEquals(mockParticipantInvited?.info.name, name)
            }
            try ifKeyExists(arg: participantInvited, key: "avatarUrl") { avatarUrl in
                try nativeAssertEquals(mockParticipantInvited?.info.avatarURL, avatarUrl)
            }
            try ifKeyExists(arg: participantInvited, key: "externalId") { externalId in
                try nativeAssertEquals(mockParticipantInvited?.info.externalID, externalId)
            }
            try ifKeyExists(arg: participantInvited, key: "permission") { permission in
                try nativeAssertEquals(mockParticipantInvited?.permissions?.first?.rawValue, permission)
            }
        }
    }

    private func assertDeclineArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.notification.declineHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }

        let mockArgs = VoxeetSDK.shared.notification.declineArgs
        try ifKeyExists(arg: args, key: "conference") { args in
            try ConferenceServiceAssertUtils.assertConference(
                args: args, mockConference: mockArgs
            )
        }
    }
}

extension NotificationServiceAsserts: SDKAsserts {

    public func runAssert(_ assert: String, args: [String: Any], result: (SDKAssertResult) -> Void) {
        do {
            switch assert {

            case "assertSubscribeArgs":
                try assertSubscribeArgs(args: args)

            case "assertUnsubscribeArgs":
                try assertUnsubscribeArgs(args: args)

            case "assertInviteArgs":
                try assertInviteArgs(args: args)

            case "assertDeclineArgs":
                try assertDeclineArgs(args: args)

            default:
                throw SDKAssertError.unknownAssert
            }
            result(.success(()))
        } catch {
            result(SDKAssertError.toResult(error: error))
        }
    }
}

